### ISSUE TITLE
Reimplement update function

### ISSUE THREAD

**Issue by 2mac** [deavmi/Gitsy#4]
`os.system()` is sloppy. Please learn to use the subprocess module instead and command line options for Git instead of using `cd` which is equally sloppy.

If you need help, I can give pointers, but I don't want to rewrite your application.

**Comment by deavmi** [comment]
I can use subprocess.call([bin,args])

**Comment by deavmi** [comment]
Should i then use what I saw somewhere. Like `git pull -C directory`

**Comment by deavmi** [comment]
I should switch to `subprocess.call` because I think, according to the Python documentation, `os.system` is going to be removed in the future.

**Comment by 2mac** [comment]
`subprocess.call` is the way to go. Yes, I do believe the `-C` option is what you are looking for.

Changing to `subprocess` will clean up your code, give you proper error codes if `git` fails for some reason, and let you control the direction of `stdout` and `stderr`.

**Comment by deavmi** [comment]
What is stdout and stderr. Is it the output from the binary printing to the screen and is stderr the rror code it has passed to you. 0 being good and others being bad in reason.

**Comment by Walkman100** [comment]
http://42ecdc9efc3e53db8fe7-76298141b7e74d4add8f8bb8e7a50188.r8.cf1.rackcdn.com/unix_admin/mobile/unix_admin_mod_3.3_SpecialCharacters-Part3_3744.mp4

**Comment by Walkman100** [comment]
^ All about std in/out/err and how to separate them

**Comment by 2mac** [comment]
`stdout` is where `print()` prints in Python (although Python has its own `stdout` that isn't completely relevant). `stderr` is an alternative stream that (usually) prints to the console, but is separate from `stdout` for scripting reasons. Essentially if I did `$ mycommand > myfile.log`, `stdout` would print to the log, but `stderr` would still print to the console to warn me of errors.

**Comment by deavmi** [comment]
Going to be updating to subprocess today.

Also. Could I do.

`from subprocess import call`

Or not?

**Comment by 2mac** [comment]
Yes. It's efficient to only import the methods you need from a module. I don't mean this as offensive, but you might take a better look at Python as a language. It seems you're unfamiliar with a lot of its syntax and standard library, and also you created a memory leak which I fixed in the pull request you merged, and I didn't think memory leaks were possible in Python.

**Comment by deavmi** [comment]
Not offensive at all. Thanks man. So to stop memory leaks should I close the file reader. `writer.close()`.

**Comment by 2mac** [comment]
Well, I changed the function to use a `with` statement, which automatically closes the open file at the end of the code block, fixing the leak.

**Comment by deavmi** [comment]
Could you also do the variable way when you make a variable into an object (the is library for writing files etc.) And then just close it at the end. Just saying. But your method is better. :D

**Comment by 2mac** [comment]
You certainly could, but your previous code did not store it in a variable.

**Comment by deavmi** [comment]
@2mac Because I didn't write to any file? Oh, for the reading of the Gitsy conf file, yes, it was just a random `open().readlines()`. :smile:

**Comment by deavmi** [comment]
@2mac Will be updating to the `git pull -C /var/www/repo` command soon, if < that is how you would go by sending a git command to a specific directory that is a git repository. ?

**Comment by 2mac** [comment]
I'd use BSD syntax to be safe. `git -C /foo/bar pull`.

**Comment by deavmi** [comment]
Okat

**Comment by 2mac** [comment]
The reason for that is not necessarily for BSD users. The options work the same on their system if it's the same program.

The issue is that in `git`, you enter commands within your commands, like `git pull`: first command `git`, second `pull`. You want to make sure the option goes to the "command" you want. putting the `-C /foo/bar` before the `pull` makes sure that `-C` is parsed as an option to `git`, since `git pull` has its own set of options. For instance `git -u pull origin master` would fail, but `git pull -u origin master` works, because `-u` is an option for `git pull` that updates `HEAD` in your repo, but it is not a recognized option for `git`.

**Comment by deavmi** [comment]
Wait, is `git pull` a seperate binary, and `git` also a seperate binary. That would make sense on machines with low resources.

**Comment by 2mac** [comment]
No. The first word (before a space) entered in the shell is the binary.

**Comment by deavmi** [comment]
Okay. :smile: @2mac

**Comment by deavmi** [comment]
https://github.com/deavmi/Gitsy/blob/master/gitsy.py#L75

How does having two variables, `o` and `a` work for this?

**Comment by 2mac** [comment]
`o` is the option string; `a` is the argument passed to it.

For example, if I write `gitsy -r myfile.txt`, `o` is `-r`, and `a` is `myfile.txt`.

**Comment by deavmi** [comment]
Woah, just discovvered something, you can put lists, into lists, dayum son.

I be like:

an_list[0][][][][][][][][][][][][][][][][][][]

for days :smile:

**Comment by deavmi** [comment]
There was a line of code that useda format string, but i think it won't work.

If i do this:

````Python
age = 2
name = "Tristan"
print("My name is %s and I am %d years old." %(name, age))
````

Nvm. Just tested it, so the use of the type of format indicator, `%s`, %d` (and whatever a floating point/decimal is), relies on what you want to do with the variable and not where it is being used in.

I would of done this, and used **%s** for the whole concantenation.


````Python
age = 2
name = "Tristan"
print("My name is %s and I am %s years old." %(name, age))
````
